pageserver/src/aux_file.rs

```diff
@@ -0,0 +1,100 @@
+use pageserver_api::key::{Key, AUX_KEY_PREFIX, METADATA_KEY_SIZE};
+use tracing::warn;
+
+/// Create a metadata key from a hash, encoded as [AUX_KEY_PREFIX, 2B directory prefix, first 13B of 128b xxhash].
+fn aux_hash_to_metadata_key(prefix: u16, data: &[u8]) -> [u8; METADATA_KEY_SIZE] {
+    let mut key = [0; METADATA_KEY_SIZE];
+    let hash = twox_hash::xxh3::hash128(data).to_be_bytes();
+    let prefix = prefix.to_be_bytes();
+    key[0] = AUX_KEY_PREFIX;
+    key[1] = prefix[0];
+    key[2] = prefix[1];
+    key[3..16].copy_from_slice(&hash[0..13]);
+    key
+}
+
+/// Encode the aux file into a fixed-size key.
+///
+/// The first byte is the AUX key prefix. We use the next 2 bytes of the key for the directory / aux file type.
+/// We have one-to-one mapping for each of the aux file that we support. We hash the remaining part of the path
+/// (usually a single file name, or several components) into 13-byte hash. The way we determine the 2-byte prefix
+/// is roughly based on the first two components of the path, one unique number for one component.
+///
+/// * pg_logical/mappings -> 0x0101
```

Comment:
Could add a case to make us a bit more robust against unexpected new sub-dirs in pg_logical:

```
/// * pg_logical/<unknown key> -> 0x01ff
```

That way if there is some new path pg_logical/foo, we won't put it into the big "others" bucket, and if we want to sort/search based on the top dir prefix in future we'll still be able to.

Reply:
fixed in https://github.com/neondatabase/neon/pull/7401/commits/6fc849b7442602b5fbd061bb48d34a624889b21d